### PR TITLE
add the implementation

### PR DESCRIPTION
**Pull request by mhlgio**
# Description

Adds a new module for obtaining embeddings from [Gradient](https://docs.gradient.ai).

> [!NOTE]
> This feature needs [Gradient Python SDK](https://pypi.org/project/gradientai/) installed, with version at least `1.1.0`.

## Type of Change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?

The change is tested with a new unit test `tests/embeddings/test_gradient.py` that obtains some embeddings for simple examples from Gradient, using the [public client](https://pypi.org/project/gradientai/). To work correctly the Gradient SDK needs to be installed:

```
pip install -U gradientai
```

The test requires the change in the fixtures to provide the access token and workspace id. They can be generated from either:

1. [Gradient website](https://auth.gradient.ai) or,
1. [Gradient CLI](https://docs.gradient.ai/docs/cli-quickstart) with `gradient env --json` command.

- [X] Added new unit/integration tests
- [X] Added new notebook (that tests end-to-end)
- [X] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/embeddings/gradient.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        except gradientai.openapi.client.exceptions.UnauthorizedException as e:
+            logger.error(f"Error while loading model {gradient_model_slug}.")
+            self._gradient.close()
+            raise ValueError("Unable to fetch the requested embeddings model") from e
+
+        super().__init__(model_name=gradient_model_slug, **kwargs)
+
+    async def _aget_query_embedding(self, query: str) -> Embedding:
+        # Gradient AI doesn't have the proper API for async yet, so we just use the sync version.
+        return self._get_query_embedding(query)
```
Comment (lilwilsond2) [inline llama_index/embeddings/gradient.py:85]: would rather just give a not yet implemented error or something

File: llama_index/embeddings/gradient.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    _model: Any = PrivateAttr()
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "GradientEmbedding"
+
+    def __init__(
+        self,
+        gradient_access_token: str,
+        gradient_workspace_id: str,
```
Comment (lilwilsond2) [inline llama_index/embeddings/gradient.py]: these should be optional so you can use your env vars

File: llama_index/embeddings/gradient.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    _gradient: Any = PrivateAttr()
+    _model: Any = PrivateAttr()
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "GradientEmbedding"
+
+    def __init__(
+        self,
```
Comment (lilwilsond2) [inline llama_index/embeddings/gradient.py:37]: can this constructor be kwargs only

File: llama_index/embeddings/__init__.py [hunk 4]
```diff
@@ -17,17 +18,18 @@
...
+    "LangchainEmbedding",
+    "LinearAdapterEmbeddingModel",
+    "OpenAIEmbedding",
     "OptimumEmbedding",
-    "resolve_embed_model",
-    "DEFAULT_HUGGINGFACE_EMBEDDING_MODEL",
     "SimilarityMode",
-    "ElasticsearchEmbeddings",
-    "ClarifaiEmbedding",
+    "resolve_embed_model",
```
Comment (lilwilsond2) [inline llama_index/embeddings/__init__.py]: that's a lot of diff

File: docs/core_modules/model_modules/embeddings/modules.md [hunk 5]
```diff
@@ -8,6 +8,8 @@ maxdepth: 1
 ---
 /examples/embeddings/OpenAI.ipynb
 /examples/embeddings/Langchain.ipynb
+/examples/embeddings/gradient_base_model.ipynb
+/examples/embeddings/gradient_model_adapter.ipynb
```
Comment (lilwilsond2) [inline docs/core_modules/model_modules/embeddings/modules.md]: the embeddings flow shouldn't really change based on us using a model adapter or base model, let's just provide one example

File: llama_index/embeddings/__init__.py (hunk 4 above)
Comment (mhlgio) [reply llama_index/embeddings/__init__.py]: sorting.

File: examples/gradient/Gradient Integration Demo.ipynb
Comment (markhng525) [inline examples/gradient/Gradient Integration Demo.ipynb:1]: The path for this cell is wrong.
`documents = SimpleDirectoryReader("../paul_graham_essay/data").load_data()`
Should be `../../docs/examples/data/paul_graham`

File: docs/examples/embeddings/gradient_base_model.ipynb
Comment (markhng525) [inline docs/examples/embeddings/gradient_base_model.ipynb:1]: The path for this cell is wrong.
`documents = SimpleDirectoryReader("../paul_graham_essay/data").load_data()`
Should be `../data/paul_graham`

File: docs/core_modules/model_modules/embeddings/modules.md (hunk 5 above)
Comment (mhlgio) [reply docs/core_modules/model_modules/embeddings/modules.md]: thought about that later, didn't update the paths. fixed.

File: llama_index/embeddings/gradient.py (hunk 3 above)
Comment (mhlgio) [reply llama_index/embeddings/gradient.py:37]: IIUC if you don't make some parameters explicit they don't show up in the IntelliSense?




 at least for me?

Comment (lilwilsond2) [reply llama_index/embeddings/gradient.py:37]: Nah I mean like `(self, *, ...)`

File: llama_index/embeddings/__init__.py (hunk 4 above)
Comment (lilwilsond2) [reply llama_index/embeddings/__init__.py]: I wouldn't mind if it was our project, but let's not do that on someone else's code base

File: llama_index/embeddings/gradient.py (hunk 1 above)
Comment (mhlgio) [reply llama_index/embeddings/gradient.py:85]: I was looking at other implementations and actually most of them implement that way :) I guess we should add async soon, tbh.

File: llama_index/embeddings/gradient.py (hunk 2 above)
Comment (mhlgio) [reply llama_index/embeddings/gradient.py]: done!

File: llama_index/embeddings/gradient.py (hunk 3 above)
Comment (mhlgio) [reply llama_index/embeddings/gradient.py:37]: THAT's what it's for!

Comment (mhlgio) [reply llama_index/embeddings/gradient.py:37]: done!

File: examples/gradient/Gradient Integration Demo.ipynb
Comment (mhlgio) [reply examples/gradient/Gradient Integration Demo.ipynb:1]: done

File: docs/examples/embeddings/gradient_base_model.ipynb
Comment (mhlgio) [reply docs/examples/embeddings/gradient_base_model.ipynb:1]: done

File: llama_index/embeddings/__init__.py (hunk 4 above)
Comment (mhlgio) [reply llama_index/embeddings/__init__.py]: fair

File: llama_index/embeddings/gradient.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    async def _aget_query_embedding(self, query: str) -> Embedding:
+        # Gradient AI doesn't have the proper API for async yet, so we just use the sync version.
+        return self._get_query_embedding(query)
+
+    def _get_text_embedding(self, text: str) -> Embedding:
+        return self._get_text_embeddings([text])[0]
+
+    def _get_text_embeddings(self, texts: List[str]) -> List[Embedding]:
+        inputs = [{"input": text} for text in texts]
```
Comment (lilwilsond2) [inline llama_index/embeddings/gradient.py:91]: since we have a limit of 100, can we batch these and return the cumulative list at the end

Comment (lilwilsond2) [reply llama_index/embeddings/gradient.py:91]: if you need a batch helper you can use something like
```python
from itertools import islice

def batched(iterable, n):
    it = iter(iterable)
    while True:
        batch = list(islice(it, n))
        if not batch:
            return
        yield batch
```

although you may be able to get some higher perf by doing something like 
```python
l = len(texts)
for i in range(0, l, n):
    yield iterable[i:(i + n)] # i'm not sure if you would need to min() with l here
```

**Pull request merged**